README.md
# plugfox_transformers  
  
[![chatroom icon](https://patrolavia.github.io/telegram-badge/chat.png)](https://t.me/PlugFox)
[![Pub](https://img.shields.io/pub/v/plugfox_transformers.svg)](https://pub.dartlang.org/packages/plugfox_transformers)  
  
## About  
  
Сontains a set of useful stream transformers
+ `Simultaneous` (Serves for simultaneous parallel tasks)
  

## Simultaneous
  
Serves for simultaneous parallel tasks

Executes simultaneously `[maxNumberOfProcesses]` generators `[convert]`
transforming each `[Event]` into a Stream of `[State]`.
The resulting stream is returned from the `[Stream.transform]` method.

If `[maxNumberOfProcesses]` is set to 0 or less, then all
incoming events processed instantly.

If `[maxNumberOfProcesses]` is set to 1, then the
behavior is almost identical .asyncExpand method.

If `[maxNumberOfProcesses]` is set to 2 or more, then this sets the number
simultaneously performed tasks and each subsequent event begins to be processed
as soon as one of the previous ones is finished.
 
The transformer breaks the sequence of events in the stream.
  

### Example usage:

```dart
Stream<String> myGenerator(int event) async* {
  String state;

  state = await Future<String>.delayed(
        const Duration(milliseconds: 250)
      , () => '$event + 2 = ${event + 2}');
  yield state;
  
  state = await Future<String>.delayed(
        const Duration(milliseconds: 500)
      , () => '$event * 2 = ${event * 2}');
  yield state;
  
  state = await Future<String>.delayed(
        const Duration(milliseconds: 750)
      , () => '$event ^ 2 = ${event * event}');
  yield state;
}

Stream<int>.fromIterable(const <int>[1, 2, 3, 4, 5, 6, 7])
  .transform<String>(Simultaneous<int, String>(myGenerator, maxNumberOfProcesses: 2))
  .forEach(print);
```

### Override `bloc` package behavior:  
  
```dart
@override  
Stream<Transition<Event, State>> transformEvents(
  Stream<Event> events,
  Stream<Transition<Event, State>> Function(Event) next) => 
    events.transform<Transition<Event, State>>(
      Simultaneous<Event, Transition<Event, State>>(next, maxNumberOfProcesses: 0)
    );
```
  
  
## Changelog  
  
Refer to the [Changelog](https://github.com/plugfox/plugfox_transformers/blob/master/CHANGELOG.md) to get all release notes.  
  
  


CHANGELOG.md
# 1.1.1  
##### 2020-04-25  
  
Disable most dev_dependencies.  
  
  
# 1.1.0  
##### 2020-04-22  
  
Update bloc example
  
  
# 1.0.2
##### 2020-04-20  
  
Update format and add example
  
  
# 1.0.1
##### 2020-04-20  
  
Update documentation for `Simultaneous` transformer
  
  
# 1.0.0
##### 2020-04-20  
  
Initial Version of the library.  
  
- Includes the ability to create a Simultaneous transformer by extending `Simultaneous` class.  
  

test/src/simultaneous_test.dart
import 'dart:async';
import 'package:test/test.dart';
import 'package:plugfox_transformers/plugfox_transformers.dart'
    show Simultaneous;

const int _delayedOperationInMilliseconds = 100;
const int _toleranceInMilliseconds = 25;
const List<int> _sourceList = <int>[1, 2, 3, 4, 5, 6, 7];
Stream<int> get _sourceStream => Stream<int>.fromIterable(_sourceList);

// ignore: unused_element
const Timeout _smallTimeout = Timeout(Duration(seconds: 1));
// ignore: unused_element
const Timeout _normalTimeout = Timeout(Duration(seconds: 15));
// ignore: unused_element
const Timeout _bigTimeout = Timeout(Duration(seconds: 360));

Stream<T> _pong<T>(T value) => Stream<T>.value(value);
Stream<T> _pongDelayed<T>(T value) => Future<T>.delayed(
    const Duration(milliseconds: _delayedOperationInMilliseconds),
    () => value).asStream();

void main() {
  test('shouldRun', () async {
    const StreamTransformer<int, int> simultaneous =
        Simultaneous<int, int>(_pong);
    await _sourceStream.transform<int>(simultaneous).drain<int>();
    expect(true, true);
  }, timeout: _smallTimeout);

  test('shouldRunAsDelayed', () async {
    const StreamTransformer<int, int> simultaneous =
        Simultaneous<int, int>(_pongDelayed);
    await _sourceStream.transform<int>(simultaneous).drain<int>();
    expect(true, true);
  }, timeout: _smallTimeout);

  test('shouldReturnExpectedData', () async {
    const StreamTransformer<int, int> simultaneous =
        Simultaneous<int, int>(_pong);
    List<int> result =
        await _sourceStream.transform<int>(simultaneous).toList();
    expect(result, _sourceList);
  }, timeout: _smallTimeout);

  test('shouldReturnExpectedDataAsDelayed', () async {
    const StreamTransformer<int, int> simultaneous =
        Simultaneous<int, int>(_pongDelayed);
    List<int> result =
        await _sourceStream.transform<int>(simultaneous).toList();
    expect(result, _sourceList);
  }, timeout: _smallTimeout);

  test('shouldBeDelayedAsUnlimited', () async {
    final int begin = DateTime.now().millisecondsSinceEpoch;
    const StreamTransformer<int, int> simultaneous =
        Simultaneous<int, int>.unlimited(_pongDelayed);
    await _sourceStream.transform<int>(simultaneous).drain<int>();
    final int end = DateTime.now().millisecondsSinceEpoch;
    final int dt = end - begin;
    expect(dt >= _delayedOperationInMilliseconds, true);
    expect(
        dt < _delayedOperationInMilliseconds + _toleranceInMilliseconds, true);
  }, timeout: _smallTimeout);

  test('shouldBeDelayedAsSingle', () async {
    final int begin = DateTime.now().millisecondsSinceEpoch;
    const StreamTransformer<int, int> simultaneous =
        Simultaneous<int, int>.single(_pongDelayed);
    await _sourceStream.transform<int>(simultaneous).drain<int>();
    final int end = DateTime.now().millisecondsSinceEpoch;
    final int dt = end - begin;
    expect(dt >= _delayedOperationInMilliseconds * _sourceList.length, true);
    expect(
        dt <
            _delayedOperationInMilliseconds * _sourceList.length +
                _toleranceInMilliseconds,
        true);
  }, timeout: _smallTimeout);

  test('shouldBeDelayedAsPair', () async {
    final int begin = DateTime.now().millisecondsSinceEpoch;
    const StreamTransformer<int, int> simultaneous =
        Simultaneous<int, int>.pair(_pongDelayed);
    await _sourceStream.transform<int>(simultaneous).drain<int>();
    final int end = DateTime.now().millisecondsSinceEpoch;
    final int dt = end - begin;
    expect(
        dt >= _delayedOperationInMilliseconds * (_sourceList.length / 2).ceil(),
        true);
    expect(
        dt <
            _delayedOperationInMilliseconds * (_sourceList.length / 2).ceil() +
                _toleranceInMilliseconds,
        true);
  }, timeout: _smallTimeout);

  test('shouldBeDelayedAsCustom', () async {
    final int begin = DateTime.now().millisecondsSinceEpoch;
    const int maxNumberOfProcesses = 5;
    const StreamTransformer<int, int> simultaneous = Simultaneous<int, int>(
        _pongDelayed,
        maxNumberOfProcesses: maxNumberOfProcesses);
    await _sourceStream.transform<int>(simultaneous).drain<int>();
    final int end = DateTime.now().millisecondsSinceEpoch;
    final int dt = end - begin;
    expect(
        dt >=
            _delayedOperationInMilliseconds *
                (_sourceList.length / maxNumberOfProcesses).ceil(),
        true);
    expect(
        dt <
            _delayedOperationInMilliseconds *
                    (_sourceList.length / maxNumberOfProcesses).ceil() +
                _toleranceInMilliseconds,
        true);
  }, timeout: _smallTimeout);

  test('reusable', () async {
    const StreamTransformer<int, int> simultaneous =
        Simultaneous<int, int>(_pong);
    List<int> result1 =
        await _sourceStream.transform<int>(simultaneous).toList();
    List<int> result2 =
        await _sourceStream.transform<int>(simultaneous).toList();
    expect(result1, result2);
  }, timeout: _smallTimeout);

  test('asSingleSubscriptionStream', () async {
    const StreamTransformer<int, int> simultaneous =
        Simultaneous<int, int>(_pong);
    expect(_sourceStream.transform<int>(simultaneous).isBroadcast, false);
  }, timeout: _smallTimeout);

  test('asBroadcastStream', () async {
    const StreamTransformer<int, int> simultaneous =
        Simultaneous<int, int>(_pong);
    expect(
        _sourceStream
            .asBroadcastStream()
            .transform<int>(simultaneous)
            .isBroadcast,
        true);
  }, timeout: _smallTimeout);

  test('cancelOnErrorFalse', () async {
    const StreamTransformer<int, int> simultaneous =
        Simultaneous<int, int>(_pong);
    bool hasError = false;
    await Stream<int>.error(Exception('test'))
        .transform<int>(simultaneous)
        .handleError((dynamic _) => hasError = true)
        .drain<int>();
    expect(hasError, true);
  }, timeout: _smallTimeout);

  test('cancelOnErrorTrue', () async {
    const StreamTransformer<int, int> simultaneous =
        Simultaneous<int, int>(_pong, cancelOnError: true);
    bool hasError = false;
    await Stream<int>.error(Exception('test'))
        .transform<int>(simultaneous)
        .handleError((dynamic _) => hasError = true)
        .drain<int>();
    expect(hasError, true);
  }, timeout: _smallTimeout);

  test('emptySourceStream', () async {
    const StreamTransformer<void, void> simultaneous =
        Simultaneous<void, void>(_pong);
    await const Stream<void>.empty()
        .transform<void>(simultaneous)
        .drain<void>();
    expect(true, true);
  }, timeout: _smallTimeout);
}


test/plugfox_transformers_test.dart
library test.plugfox_transformers;

import 'package:test/test.dart';

import 'src/simultaneous_test.dart' as simultaneous_test;

void main() {
  group('plugfox_transformers.Simultaneous', simultaneous_test.main);
}


lib/src/simultaneous.dart
import 'dart:async';

/// Serves for simultaneous parallel tasks
///
/// Executes simultaneously [maxNumberOfProcesses] generators [convert]
/// transforming each [Event] into a Stream of [State].
/// The resulting stream is returned from the [Stream.transform] method.
///
/// If [maxNumberOfProcesses] is set to 0 or less, then all
/// incoming events processed instantly.
///
/// If [maxNumberOfProcesses] is set to 1, then the
/// behavior is almost identical .asyncExpand method.
///
/// If [maxNumberOfProcesses] is set to 2 or more, then this sets the number
/// simultaneously performed tasks and each subsequent event begins to be processed
/// as soon as one of the previous ones is finished.
///
/// The transformer breaks the sequence of events in the stream.
///
///
/// **Usage example:**
/// ```
///   events.transform<State>(Simultaneous<Event, State>(myGenerator, maxNumberOfProcesses: 2))
/// ```
///      // source:              abcd------------------
///      // source.delayed(2):   --A1B1A2B2--C1D1C2D2--
///
///
/// **Example usage to override the behavior of the [bloc] library:**
/// ```
///   @override
///   Stream<Transition<Event, State>> transformEvents(
///     Stream<Event> events,
///     Stream<Transition<Event, State>> Function(Event) next) =>
///       events.transform<Transition<Event, State>>(
///         Simultaneous<Event, Transition<Event, State>>(next, maxNumberOfProcesses: 0)
///       );
/// ```
///
class Simultaneous<Event, State> implements StreamTransformer<Event, State> {
  final Stream<State> Function(Event) _convert;
  final int _maxProcesses;
  final bool _cancelOnError;

  /*
  /// Set maximum number of simultaneous parallel tasks
  set maxNumberOfProcesses(int value) =>
    _maxProcesses = value ?? 0;
  */

  /// Transformer with custom maximum number of concurent processes
  const Simultaneous(Stream<State> Function(Event) convert,
      {int maxNumberOfProcesses = 0, bool cancelOnError = false})
      : assert(convert != null && maxNumberOfProcesses != null),
        _convert = convert,
        _maxProcesses = maxNumberOfProcesses ?? 0,
        _cancelOnError = cancelOnError ?? false;

  /// Transformer with an infinite number of concurent processes
  const Simultaneous.unlimited(Stream<State> Function(Event) convert,
      {bool cancelOnError = false})
      : assert(convert != null),
        _convert = convert,
        _maxProcesses = 0,
        _cancelOnError = cancelOnError ?? false;

  /// Transformer with one concurent process
  const Simultaneous.single(Stream<State> Function(Event) convert,
      {bool cancelOnError = false})
      : assert(convert != null),
        _convert = convert,
        _maxProcesses = 1,
        _cancelOnError = cancelOnError ?? false;

  /// Transformer with two concurent processes
  const Simultaneous.pair(Stream<State> Function(Event) convert,
      {bool cancelOnError = false})
      : assert(convert != null),
        _convert = convert,
        _maxProcesses = 2,
        _cancelOnError = cancelOnError ?? false;

  @override
  Stream<State> bind(Stream<Event> events) {
    final _SimultaneousController controller =
        _SimultaneousController(maxNumberOfTasks: _maxProcesses);
    final StreamIterator<Event> eventIterator = StreamIterator<Event>(events);
    final StreamController<State> stateController = StreamController<State>(
      onPause: null,
      onResume: null,
      onCancel: eventIterator.cancel,
      sync: false,
    );
    // ITERATOR +
    Future.doWhile(
      () => eventIterator.moveNext().then((bool hasNext) async {
        if (!hasNext) {
          await controller.awaitLastTask();
          return await Future.wait<void>(<Future<void>>[
            controller.close(),
            stateController.close(),
          ]).then<bool>((void _) => false);
        }
        Future<void> callback = _convert(eventIterator.current)
            .handleError((dynamic error, StackTrace stackTrace) {
          stateController.addError(error, stackTrace);
          if (_cancelOnError) {
            eventIterator.cancel();
            controller.close();
            stateController.close();
          }
        }).forEach(stateController.sink.add);
        await controller.awaitTaskIfNeeded(callback);
        return true;
      }, onError: (dynamic error, StackTrace stackTrace) {
        stateController.addError(error, stackTrace);
        if (_cancelOnError) {
          eventIterator.cancel();
          controller.close();
          stateController.close();
          return false;
        }
        return true;
      }),
    );
    // ITERATOR -
    return events.isBroadcast
        ? stateController.stream.asBroadcastStream()
        : stateController.stream;
  }

  @override
  StreamTransformer<RS, RT> cast<RS, RT>() =>
      StreamTransformer.castFrom<Event, State, RS, RT>(this);
}

class _SimultaneousController {
  int _currTasks;
  final int _maxTasks;
  final StreamController<int> _currentTasksChangedController =
      StreamController<int>.broadcast();
  bool get hasTask => _currTasks > 0;
  bool get awaitTask => _currTasks >= _maxTasks && _maxTasks > 0;

  _SimultaneousController({int maxNumberOfTasks = 0})
      : assert(maxNumberOfTasks != null),
        _maxTasks = maxNumberOfTasks ?? 0,
        _currTasks = 0;

  FutureOr<void> awaitTaskIfNeeded(Future<void> callback) {
    _addTask();
    // ignore: unawaited
    callback.whenComplete(_rmTask);
    if (awaitTask) return callback;
  }

  void _addTask() {
    _currTasks++;
    _currentTasksChangedController.add(_currTasks);
  }

  void _rmTask() {
    _currTasks--;
    _currentTasksChangedController.add(_currTasks);
  }

  FutureOr<void> awaitLastTask() async {
    if (_currTasks != 0) {
      await _currentTasksChangedController.stream.firstWhere((int n) => n == 0);
    }
  }

  Future<void> close() => _currentTasksChangedController.close();
}


example/main.dart
// ignore_for_file: avoid_print

import 'package:plugfox_transformers/plugfox_transformers.dart';

void main() {
  // Simultaneous Transformer
  Stream<String> myGenerator(int event) async* {
    String state;

    state = await Future<String>.delayed(
        const Duration(milliseconds: 250), () => '$event + 2 = ${event + 2}');
    yield state;

    state = await Future<String>.delayed(
        const Duration(milliseconds: 500), () => '$event * 2 = ${event * 2}');
    yield state;

    state = await Future<String>.delayed(const Duration(milliseconds: 750),
        () => '$event ^ 2 = ${event * event}');
    yield state;
  }

  Stream<int>.fromIterable(const <int>[1, 2, 3, 4, 5, 6, 7])
      .transform<String>(
          Simultaneous<int, String>(myGenerator, maxNumberOfProcesses: 2))
      .forEach(print);
}
